python/ray/experimental/gpu_object_manager/gpu_object_store.py

```diff
@@ -78,36 +79,96 @@ def __ray_recv__(
         tensor = torch.zeros(shape, dtype=dtype, device=device)
         collective.recv(tensor, src_rank, group_name=communicator_name)
         tensors.append(tensor)
-    gpu_object_store.add_gpu_object(obj_id, tensors)
+    gpu_object_store.add_object(obj_id, tensors)
+
+
+def __ray_get_tensor_meta__(self, obj_id: str):
+    """Helper function that runs on the src actor to get the tensor metadata."""
+    from ray._private.worker import global_worker
+
+    gpu_object_store = global_worker.gpu_object_manager.gpu_object_store
+    # NOTE: We do not specify a timeout here because the user task that returns
+    # it could take arbitrarily long and we don't want to trigger a spurious
+    # timeout.
+    gpu_object_store.wait_object(obj_id)
+    tensors = gpu_object_store.get_object(obj_id)
+    return [(t.shape, t.dtype) for t in tensors]
 
 
 def __ray_fetch_gpu_object__(self, obj_id: str):
     """Helper function that runs on the src actor to fetch tensors from the GPU object store via the object store."""
     from ray._private.worker import global_worker
 
     gpu_object_store = global_worker.gpu_object_manager.gpu_object_store
-    assert gpu_object_store.has_gpu_object(
+    assert gpu_object_store.has_object(
         obj_id
     ), f"obj_id={obj_id} not found in GPU object store"
-    tensors = gpu_object_store.get_gpu_object(obj_id)
+    tensors = gpu_object_store.get_object(obj_id)
     return tensors
 
 
 class GPUObjectStore:
-    def __init__(self):
+    def __init__(self, gpu_object_store_lock: threading.RLock):
         # A dictionary that maps from an object ID to a list of tensors.
         #
         # Note: Currently, `gpu_object_store` is only supported for Ray Actors.
         self.gpu_object_store: Dict[str, List["torch.Tensor"]] = {}
+        # Synchronization for GPU object store. The GPU object store can be read
+        # and written by the main thread, which is executing user code, and the
+        # background _ray_system thread, which executes data transfers and
+        # performs GC.
+        self.object_present_cv = threading.Condition(gpu_object_store_lock)
 
-    def has_gpu_object(self, obj_id: str) -> bool:
+    def has_object(self, obj_id: str) -> bool:
         return obj_id in self.gpu_object_store
 
-    def get_gpu_object(self, obj_id: str) -> Optional[List["torch.Tensor"]]:
+    def get_object(self, obj_id: str) -> Optional[List["torch.Tensor"]]:
         return self.gpu_object_store[obj_id]
```

Comment:
don't these need to be guarded by the lock?

Reply:
Python dictionary operations are atomic.